src/eepromfs.c

```diff
@@ -0,0 +1,888 @@
+    {
+        /* No bytes to read; bail early */
+        return 0;
+    }
+
+    /* Calculate the starting point to read out of EEPROM from */
+    size_t current_block = file->start_block + (cursor / EEPROM_BLOCK_SIZE);
+    size_t block_byte_offset = cursor % EEPROM_BLOCK_SIZE;
+    int bytes_read = 0;
+
+    uint8_t eeprom_buf[EEPROM_BLOCK_SIZE];
+    uint8_t * ptr = dest;
+
+    do {
+        eeprom_read(current_block, eeprom_buf);
+        /* Fill the buffer with the data from the current block */
+        while ( bytes_to_read > 0 && block_byte_offset < EEPROM_BLOCK_SIZE )
+        {
+            *(ptr++) = eeprom_buf[block_byte_offset];
+            bytes_read++;
+            bytes_to_read--;
+            block_byte_offset++;
+        }
+        /* Move on to the next block */
+        block_byte_offset = 0;
+        current_block++;
+    } while ( bytes_to_read > 0 );
+
+    file->cursor = cursor + bytes_read;
+
+    return bytes_read;
+}
+
+/**
+ * @brief Writes a specific amount of data to a file starting from the cursor.
+ *
+ * @param[in] path
+ *            Path of file in EEPROM filesystem to write to
+ * @param[in] src
+ *            Buffer containing the data to write
+ * @param[in] len
+ *            Number of bytes to write
+ *
+ * @return The actual number of bytes written or a negative value on failure.
+ */
+static int eepfs_cursor_write(eepfs_file_t * const file, const void * const src, const size_t len)
+{
+    if ( file == NULL )
+    {
+        return EEPFS_EBADHANDLE;
+    }
+
+    if ( src == NULL ) 
+    {
+        return EEPFS_EBADINPUT;
+    }
+
+    size_t cursor = file->cursor;
+    size_t bytes_to_write = len;
+
+    /* Bounds check to make sure it doesn't write past the end */
+    if ( cursor + bytes_to_write >= file->max_cursor )
+    {
+        /* It will, let's shorten it */
+        bytes_to_write = file->max_cursor - cursor;
+    }
+
+    if ( bytes_to_write <= 0 )
+    {
+        /* No bytes to write; bail early */
+        return 0;
+    }
+
+    /* Calculate the starting point to write to EEPROM from */
+    size_t current_block = file->start_block + (cursor / EEPROM_BLOCK_SIZE);
+    size_t block_byte_offset = cursor % EEPROM_BLOCK_SIZE;
+    int bytes_written = 0;
+
+    uint8_t eeprom_buf[EEPROM_BLOCK_SIZE];
+    const uint8_t * ptr = src;
+
+    do {
+        /* Only read in the current block if we need to preserve data from it */
+        if ( bytes_to_write < EEPROM_BLOCK_SIZE || block_byte_offset != 0 )
+        {
+            eeprom_read(current_block, eeprom_buf);
+        }
+        /* Fill the current block with the data to write */
+        while ( bytes_to_write > 0 && block_byte_offset < EEPROM_BLOCK_SIZE )
+        {
+            eeprom_buf[block_byte_offset] = *(ptr++);
+            bytes_written++;
+            bytes_to_write--;
+            block_byte_offset++;
+        }
+        /* Write the block and move on to the next one */
+        eeprom_write(current_block, eeprom_buf);
```

Comment:
Given the above comment on caching the current block, I would defer actual writing to when the user has issued a write commond on a different block, so that each block is normally written only once.

Reply:
Then we may also need a flush function to gracefully write pending writes in order not to loose information IMO.